www/js/diary/addressNamesHelper.ts

```diff
@@ -0,0 +1,166 @@
+/* This is a temporary solution; localstorage is not a good long-term option and we should
+  be looking to other key-value storage options in the React Native ecosystem. */
+
+import { useEffect, useState, useRef, useContext } from 'react';
+
+export type Listener<EventType> = (event: EventType) => void;
+
+export type ObserverReturnType<KeyType, EventType> = {
+  subscribe: (entryKey: KeyType, listener: Listener<EventType>) => () => void;
+  publish: (entryKey: KeyType, event: EventType) => void;
+};
+
+export default function createObserver<
+  KeyType extends string | number | symbol,
+  EventType,
+>(): ObserverReturnType<KeyType, EventType> {
+  const listeners: Record<KeyType, Listener<EventType>[]> = {} as Record<
+    KeyType,
+    Listener<EventType>[]
+  >;
+
+  return {
+    subscribe: (entryKey: KeyType, listener: Listener<EventType>) => {
+      if (!listeners[entryKey]) listeners[entryKey] = [];
+      listeners[entryKey].push(listener);
+      return () => {
+        listeners[entryKey].splice(listeners[entryKey].indexOf(listener), 1);
+      };
+    },
+    publish: (entryKey: KeyType, event: EventType) => {
+      if (!listeners[entryKey]) listeners[entryKey] = [];
+      listeners[entryKey].forEach((listener: Listener<EventType>) =>
+        listener(event),
+      );
+    },
+  };
+}
+
+export const LocalStorageObserver = createObserver<string, string>();
+
+export const { subscribe, publish } = LocalStorageObserver;
+
+export function useLocalStorage<T>(key: string, initialValue: T) {
+
+  const [storedValue, setStoredValue] = useState(() => {
+    try {
+      const item = window.localStorage.getItem(key);
+      return item || initialValue;
+    } catch (error) {
+      return initialValue;
+    }
+  });
+
+  const keyRef = useRef(key);
+  if (keyRef.current !== key) {
+    keyRef.current = key;
+    // force state update
+    const storedValue = window.localStorage.getItem(key);
+    setStoredValue(storedValue);
+  }
+
+  LocalStorageObserver.subscribe(key, setStoredValue);
+
+  const setValue = (value: T) => {
+    try {
+      const valueToStore =
+        value instanceof Function ? value(storedValue) : value;
+      setStoredValue(valueToStore);
+      LocalStorageObserver.publish(key, valueToStore);
+      if (typeof window !== 'undefined') {
+        window.localStorage.setItem(key, JSON.stringify(valueToStore));
+      }
+    } catch (error) {
+      console.error(error);
```

Comment:
nit: I don't like errors that are silently swallowed - if this happens in the field, we have no visibility into it either through alert or through logs. If we anticipate that this is unlikely, particularly since it is related to somewhat flaky storage, we might want to use `Logger.displayError` instead.